wm/client.go

```diff
@@ -0,0 +1,258 @@
+// +build linux,!ci
+
+package wm
+
+import (
+	"github.com/BurntSushi/xgb/xproto"
+	"github.com/BurntSushi/xgbutil/ewmh"
+	"github.com/BurntSushi/xgbutil/icccm"
+	"github.com/BurntSushi/xgbutil/xevent"
+	"github.com/BurntSushi/xgbutil/xprop"
+
+	"fyne.io/desktop"
+	"fyne.io/fyne"
+)
+
+type client struct {
+	id, win xproto.Window
+
+	framed    bool
+	iconic    bool
+	maximized bool
+	title     string
+	class     []string
+	command   string
+	iconName  string
+
+	allowedActions []string
+	stateHints     []string
+
+	fr *frame
+	wm *x11WM
+}
+
+func (s *stack) clientForWin(id xproto.Window) desktop.Window {
+	for _, w := range s.clients {
+		if w.(*client).id == id || w.(*client).win == id {
+			return w
+		}
+	}
+
+	return nil
+}
+
+func (cli *client) Decorated() bool {
+	return cli.framed
+}
+
+func (cli *client) Title() string {
+	if cli.title == "" {
+		cli.title = windowName(cli.wm.x, cli.win)
+	}
+	return cli.title
+}
+
+func (cli *client) Class() []string {
+	if len(cli.class) == 0 {
+		cli.class = windowClass(cli.wm.x, cli.win)
+	}
+	return cli.class
+}
+
+func (cli *client) Command() string {
+	if cli.command == "" {
+		cli.command = windowCommand(cli.wm.x, cli.win)
+	}
+	return cli.command
+}
+
+func (cli *client) IconName() string {
+	if cli.iconName == "" {
+		cli.iconName = windowIconName(cli.wm.x, cli.win)
+	}
+	return cli.iconName
+}
+
+func (cli *client) Iconic() bool {
+	return cli.iconic
+}
+
+func (cli *client) Maximized() bool {
+	return cli.maximized
+}
+
+func (cli *client) TopWindow() bool {
+	if cli.wm.TopWindow() == cli {
+		return true
+	}
+	return false
+}
+
+func (cli *client) Close() {
+	winProtos, err := icccm.WmProtocolsGet(cli.wm.x, cli.win)
+	if err != nil {
+		fyne.LogError("Get Protocols Error", err)
+	}
+
+	askNicely := false
+	for _, proto := range winProtos {
+		if proto == "WM_DELETE_WINDOW" {
+			askNicely = true
+		}
+	}
+
+	if !askNicely {
+		err := xproto.DestroyWindowChecked(cli.wm.x.Conn(), cli.win).Check()
+		if err != nil {
+			fyne.LogError("Close Error", err)
+		}
+
+		return
+	}
+
+	protocols, err := xprop.Atm(cli.wm.x, "WM_PROTOCOLS")
+	if err != nil {
+		fyne.LogError("Get Protocols Error", err)
+		return
+	}
+
+	delWin, err := xprop.Atm(cli.wm.x, "WM_DELETE_WINDOW")
+	if err != nil {
+		fyne.LogError("Get Delete Window Error", err)
+		return
+	}
+	cm, err := xevent.NewClientMessage(32, cli.win, protocols,
+		int(delWin))
+	err = xproto.SendEventChecked(cli.wm.x.Conn(), false, cli.win, 0,
+		string(cm.Bytes())).Check()
+	if err != nil {
+		fyne.LogError("Window Delete Error", err)
+	}
+}
+
+func (cli *client) iconifyMessage(state int) {
```

Comment:
This is not really iconifyMessage - it is actually "sendMessage".
the "iconifyAtm" should just be "stateChangeAtom" and it can be used by other parts of code.